gpu_bdb/queries/q03/gpu_bdb_query_03_dask_sql.py

```diff
@@ -0,0 +1,232 @@
+# Copyright (c) 2019-2020, BlazingSQL, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import sys
+import os
+
+from bdb_tools.cluster_startup import attach_to_cluster
+from numba import cuda
+
+from bdb_tools.utils import (
+    benchmark,
+    gpubdb_argparser,
+    run_query,
+)
+
+from bdb_tools.readers import build_reader
+
+from dask.distributed import wait
+
+
+# -------- Q03 -----------
+q03_days_in_sec_before_purchase = 864000
+q03_views_before_purchase = 5
+q03_purchased_item_IN = 10001
+# --see q1 for categories
+q03_purchased_item_category_IN = "2,3"
+q03_limit = 100
+
+
+@cuda.jit
+def find_items_viewed_before_purchase_kernel(
+    relevant_idx_col, user_col, timestamp_col, item_col, out_col, N
+):
+    """
+    Find the past N items viewed after a relevant purchase was made,
+    as defined by the configuration of this query.
+    """
+    i = cuda.grid(1)
+    relevant_item = q03_purchased_item_IN
+
+    if i < (relevant_idx_col.size):  # boundary guard
+        # every relevant row gets N rows in the output, so we need to map the indexes
+        # back into their position in the original array
+        orig_idx = relevant_idx_col[i]
+        current_user = user_col[orig_idx]
+
+        # look at the previous N clicks (assume sorted descending)
+        rows_to_check = N
+        remaining_rows = user_col.size - orig_idx
+
+        if remaining_rows <= rows_to_check:
+            rows_to_check = remaining_rows - 1
+
+        for k in range(1, rows_to_check + 1):
+            if current_user != user_col[orig_idx + k]:
+                out_col[i * N + k - 1] = 0
+
+            # only checking relevant purchases via the relevant_idx_col
+            elif (timestamp_col[orig_idx + k] <= timestamp_col[orig_idx]) & (
+                timestamp_col[orig_idx + k]
+                >= (timestamp_col[orig_idx] - q03_days_in_sec_before_purchase)
+            ):
+                out_col[i * N + k - 1] = item_col[orig_idx + k]
+            else:
+                out_col[i * N + k - 1] = 0
+
+def load_q03():
+    import importlib, types
+
+    fn = os.path.join(os.getcwd(), "gpu_bdb_query_03_dask_sql.py")
+    if not os.path.isfile(fn):
+        fn = os.path.join(os.getcwd(), "queries/q03/gpu_bdb_query_03_dask_sql.py")
+
+    loader = importlib.machinery.SourceFileLoader("03", fn)
+    mod = types.ModuleType(loader.name)
+    loader.exec_module(mod)
+    sys.modules[loader.name] = mod
+    return mod.main
```

Comment:
This looks weird to me . Why are we doing this again ?

Reply:
Some changes to Numba upstream (0.54+) no longer identifies the kernel if the file was imported from the benchmark runner via import lib, and the module containing the kernel needs to be explicitly added to the sys modules and loaded before being called. cc: @brandon-b-miller who can probably provide more insight on this

Reply:
Gotcha. Looks really confusing . I wonder if we can work around importing using `import lib`  if there is no work around. Will wait on @brandon-b-miller  to comment .